shared/prow/prow.go

```diff
@@ -0,0 +1,308 @@
+	Latest        = "latest-build.txt"
+	// BuildLog is the filename for build log
+	BuildLog      = "build-log.txt"
+	// StartedJSON is the json file containing build started info
+	StartedJSON   = "started.json"
+	// FinishedJSON is the json file containing build finished info
+	FinishedJSON  = "finished.json"
+	// ArtifactsDir is the dir containing artifacts
+	ArtifactsDir  = "artifacts"
+
+	// PresubmitJob means it runs on unmerged PRs.
+	PresubmitJob  = "presubmit"
+	// PostsubmitJob means it runs on each new commit.
+	PostsubmitJob = "postsubmit"
+	// PeriodicJob means it runs on a time-basis, unrelated to git changes.
+	PeriodicJob   = "periodic"
+	// BatchJob tests multiple unmerged PRs at the same time.
+	BatchJob      = "batch"
+)
+
+var ctx = context.Background()
+
+// Job struct represents a job directory in gcs.
+// gcs job StoragePath will be derived from Type if it's defined,
+type Job struct {
+	Name        string
+	Type        string
+	Bucket      string // optional
+	Repo        string // optional
+	StoragePath string // optional
+	PullID      int // only for Presubmit jobs
+	Builds      []Build // optional
+}
+
+// Build points to a build stored under a particular gcs path.
+type Build struct {
+	JobName	    string
+	StoragePath string
+	BuildID	    int
+	Bucket      string // optional
+}
+
+// Started holds the started.json values of the build.
+type Started struct {
+	Timestamp   int64             `json:"timestamp"` // epoch seconds
+	RepoVersion string            `json:"repo-version"`
+	Node        string            `json:"node"`
+	Pull        string            `json:"pull"`
+	Repos       map[string]string `json:"repos"` // {repo: branch_or_pull} map
+}
+
+// Finished holds the finished.json values of the build
+type Finished struct {
+	// Timestamp is epoch seconds
+	Timestamp   int64            `json:"timestamp"`
+	Passed      bool             `json:"passed"`
+	JobVersion  string           `json:"job-version"`
+	Metadata    Metadata         `json:"metadata"`
+}
+
+// Metadata contains metadata in finished.json
+type Metadata map[string]interface{}
+
+/* Local logics */
+
+// GetLocalArtifactsDir gets the aritfacts directory where prow looks for artifacts.
+// By default, it will look at the env var ARTIFACTS.
+func GetLocalArtifactsDir() string {
+	dir := os.Getenv("ARTIFACTS")
+	if dir == "" {
+		log.Printf("Env variable ARTIFACTS not set. Using %s instead.", ArtifactsDir)
+		dir = ArtifactsDir
+	}
+	return dir
+}
+
+/* GCS related logics */
+
+// Initialize wraps gcs authentication, have to be invoked before any other functions
+func Initialize(serviceString string) error {
+	return gcs.Authenticate(ctx, serviceString)
+}
+
+// NewJob creates new job struct
+// pullID is only saved by Presubmit job for determining StoragePath
+func NewJob(jobName, jobType, repoName string, pullID int) *Job {
+	job := Job{
+		Name: jobName,
+		Type: jobType,
+		Bucket: BucketName,
+	}
+
+	switch jobType {
```

Comment:
You're not failing when the job type is unknown anymore. StoragePath will be empty/invalid. Is this intended?

Reply:
Thought about this again, realized that job with gcs path of empty string is very dangerous, especially when implementing deleting in the future. Failed with unknown job type again, and added separate unit test for unknow job type.